Start addition of instructions for external users to README (#106)
AlexsLemonade/scpca-nf: external-data-instructions.md

  - [Prepare the metadata file](#prepare-the-metadata-file)
  - [Configuring `scpca-nf` for your environment](#configuring-scpca-nf-for-your-environment)
    - [Configuration files](#configuration-files)
    - [Setting up a profile in the configuration file](#setting-up-a-profile-in-the-configuration-file)
    - [Using `scpca-nf` with AWS](#using-scpca-nf-with-aws)
  - [Repeating the mapping step](#repeating-the-mapping-step)
  - [Special considerations for using `scpca-nf` with spatial transcriptomics libraries](#special-considerations-for-using-scpca-nf-with-spatial-transcriptomics-libraries)
  - [Output files](#output-files)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# How to use `scpca-nf` as an external user 

In order to use `scpca-nf` to process your own data, you will need to make sure you have the following installed: 

- [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html#installation)
- [Docker](https://www.nextflow.io/docs/latest/docker.html) or [Singlularity](https://www.nextflow.io/docs/latest/singularity.html) (installed either locally or on an HPC)

> You will also need to have files organized so that all the sequencing files for each run are in their own directory or folder.

Reviewer: I don't think we will have defined what a run is by the time I've gotten here, right? I think we mean file/set of files.

Any sequencing runs that contain multiple samples must be demultiplexed and FASTQ files must be placed into separate distinct folders. 

> If a run has been re-sequenced for any reason (e.g. to increase coverage), all FASTQ files should be combined into the same folder. 

Reviewer: ```suggestion
If a run has been re-sequenced for any reason (e.g., to increase coverage), all FASTQ files should be combined into the same folder. 
```

Reviewer: But I'm fairly sure we mean library here?

> 

Reviewer: Should we instead just say "you have to organize your files" and then add a section (after metadata, perhaps) about preparing your file organization?

Reviewer: Actually I guess I mean take the first part of the metadata section and split it out into its own thing.

Once you have set up your environment and created these files you will be able to start your run as follows, adding any additional optional parameters that you may choose: 

```bash
nextflow run AlexsLemonade/scpca-nf \
  -r v0.2.4 \
  -config my_config.config \
  --run_metafile <path/to/metadata_file>
```

This will pull the `scpca-nf` workflow directly from Github, using the `v0.2.4` version, and run it based on the settings in the local configuration file `my_config.config`.

**Note:** `scpca-nf` is under active development.
We strongly encourage you to use a release tagged version of the workflow, set here with the `-r` flag.
Released versions can be found on the [`scpca-nf` repo releases page](https://github.com/AlexsLemonade/scpca-nf/releases).

## Prepare the metadata file 

Using `scpca-nf` requires a metadata file as a TSV (tab separated values) file, where each sequencing run to be processed is a row and columns contain associated information about that run. 

For each run, you will need to provide a Run ID (`scpca_run_id`), library ID (`scpca_library_id`), and sample ID (`scpca_sample_id`). 

> Each row should contain a unique run ID, corresponding to a sequencing run or set of FASTQ files. 

Reviewer: The information about set of files needs to get moved up.

In most Batch queue setups, each AWS compute node has a fixed amount of disk space. 
We found it useful to have two queues: one for general use and one for jobs that may require larger amounts of disk space. 
The two compute environments use the same AMI, but use [Launch Templates](https://docs.aws.amazon.com/batch/latest/userguide/launch-templates.html) to configure the nodes on launch with different amounts of disk space. 
Currently, our default queue is configured with a disk size of 128 GB for each node, and our "bigdisk" queue has 1000 GB of disk space.
The queue used by each process is determined by Nextflow labels and associated profile settings.

The Data Lab's [AWS Batch config file](https://github.com/AlexsLemonade/scpca-nf/blob/main/config/profile_awsbatch.config) may be helpful as a reference for creating a profile for use with AWS, but note that the queues and file locations listed there are not publicly available, so these will need to be set to different values your own profile.

> ## Repeating the mapping step

Reviewer: ```suggestion
## Repeating mapping steps
```

The run ID will correspond to the name of the folder that contains the FASTQ files associated with the sequencing run.
See [the section on file organization above for more information](#file-organization).

The library ID will be unique for each set of cells that have been isolated from a sample and have undergone droplet generation. 
For single-cell/single-nuclei RNA-seq runs, the library ID should be unique for each sequencing run.
For libraries that have corresponding CITE-seq, they should share the same library ID as the associated single-cell/single-nuclei RNA-seq run, indicating that the sequencing data has been generated from the same group of cells. 

Finally, the sample ID will indicate the unique tissue or source from which a sample was collected. 
If you have two libraries that have been generated from the same original tissue, then they will share the same sample ID. 

For more information on understanding the difference between library and sample IDs, see the [FAQ on library and sample IDs in the ScPCA portal documentation](https://scpca.readthedocs.io/en/latest/faq.html#what-is-the-difference-between-samples-and-libraries). 

> Before using the workflow with data that you might plan to submit to ScPCA, please be sure to obtain a list of sample identifiers to use for your samples from the Data Lab. 

Reviewer: ```suggestion
**Before using the workflow with data that you might plan to submit to ScPCA, please be sure to obtain a list of sample identifiers to use for your samples from the Data Lab.**
```

We could first create a file `my_config.config` (or a filename of your choice) with the following contents:  

```groovy
// my_config.config
params.run_metafile = '<path/to/metadata_file>'
params.outdir = '<path/to/output>'
```

This file is then used with the `-config` (or `-c`) argument at the command line:

```sh
nextflow run AlexsLemonade/scpca-nf \
  -r v0.2.4 \
  -config my_config.config 
```

For reference, we provide an example template configuration file, [user_template.config](https://github.com/AlexsLemonade/scpca-nf/blob/main/examples/user_template.config), which includes some other workflow parameters that may be useful, as well as an example of profile configuration, discussed below. 

See the [Nextflow documentation](https://www.nextflow.io/docs/latest/config.html) and the below sections for more detail on creating your own configuration file.

### Setting up a profile in the configuration file

Local running may be sufficient for small jobs or testing, but you will most likely want to run your workflow in a high performance computing environment (HPC), such as an institutional computing cluster or on a cloud service like AWS.

> To do this, we recommend using [Nextflow profiles](https://www.nextflow.io/docs/latest/config.html#config-profiles) to encapsulate settings like the [`executor`](https://www.nextflow.io/docs/latest/executor.html) that will be used to run each process and associated details that may be required, such as queue names or the container engine (i.e. [Docker](https://www.nextflow.io/docs/latest/docker.html) or [Singularity](https://www.nextflow.io/docs/latest/singularity.html)) your system uses.

Reviewer: ```suggestion
To do this, we recommend using [Nextflow profiles](https://www.nextflow.io/docs/latest/config.html#config-profiles) to encapsulate settings like the [`executor`](https://www.nextflow.io/docs/latest/executor.html) that will be used to run each process and associated details that may be required, such as queue names or the container engine (i.e., [Docker](https://www.nextflow.io/docs/latest/docker.html) or [Singularity](https://www.nextflow.io/docs/latest/singularity.html)) your system uses.
```

- [How to use `scpca-nf` as an external user](#how-to-use-scpca-nf-as-an-external-user)
  - [Prepare the metadata file](#prepare-the-metadata-file)
  - [Configuring `scpca-nf` for your environment](#configuring-scpca-nf-for-your-environment)
    - [Setting up a profile in the config file](#setting-up-a-profile-in-the-config-file)
    - [Adding parameters to the config file](#adding-parameters-to-the-config-file)
    - [Using `scpca-nf` with AWS](#using-scpca-nf-with-aws)
  - [Adjust optional parameters](#adjust-optional-parameters)
  - [Special considerations for using `scpca-nf` with spatial transcriptomics libraries](#special-considerations-for-using-scpca-nf-with-spatial-transcriptomics-libraries)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## How to use `scpca-nf` as an external user 

In order to use `scpca-nf` to process your own data, you will need to make sure you have the following installed: 

- [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html#installation)
- [Docker](https://docs.docker.com/get-started/#download-and-install-docker) (installed either locally or on an HPC)

> Once you have setup your environment and completed the below steps you will be able to setup your run as follows, adding any additional optional parameters that you may choose: 

Reviewer 2: I want to give a sense of what the setup referred to "below" is to reduce the cognitive load. I think the wording still needs work. We might also not want to talk about parameters at this stage, as we might expect (encourage) people to set those in the config file?

```suggestion
You will also need to create a metadata file and a nextflow configuration file (see below).
Once you have set up your environment and created these files you will be able to start your run as follows, adding any additional optional parameters that you may choose: 
```

```
nextflow run AlexsLemonade/scpca-nf \

>  -r 0.2.3 \

Reviewer 2: It is probably worth adding a comment that mentions that the `-r` flag is required (or highly recommended).

 -c my_config.config \
  --run_metafile <path/to/metadata file> \
  --outdir <path/to/output>
```

### Prepare the metadata file 

Using `scpca-nf` requires a metadata file where each library to be processed is a row and columns contain associated information about that library. 
You will need the following columns to be present in your metadata file... 

### Configuring `scpca-nf` for your environment

By default, the workflow is set up to run in a local environment. 
If choosing to run the workflow locally, you will only need to provide the path to the metadata file and output directory where you would like the results to be stored. 

```
nextflow run AlexsLemonade/scpca-nf \

>  -r 0.2.3 \

Reviewer 2: I think this code snippet may be redundant?
